src/base58.cpp

```diff
@@ -84,44 +90,47 @@ static const int8_t mapBase58[256] = {
     return true;
 }
 
+auto ConvertToLongs(const Span<const unsigned char>& input, int start) -> std::vector<int64_t> {
+    int effectiveLength = input.size() - start;
+    std::vector<int64_t> inputAsLongs((effectiveLength + (encodeGroupSize - 1)) / encodeGroupSize);
+    int groupOffset = ((encodeGroupSize - (effectiveLength % encodeGroupSize)) % encodeGroupSize) - start;
+
+    for (size_t i = start; i < input.size(); ++i) {
+        int index = (groupOffset + static_cast<int>(i)) / encodeGroupSize;
+        inputAsLongs[index] <<= 8;
+        inputAsLongs[index] |= input[i];
+    }
+
+    return inputAsLongs;
+}
+
 std::string EncodeBase58(Span<const unsigned char> input)
 {
-    // Skip & count leading zeroes.
-    int zeroes = 0;
-    int length = 0;
-    while (input.size() > 0 && input[0] == 0) {
-        input = input.subspan(1);
-        zeroes++;
+    auto leading{0U};
+    while (leading < input.size() && input[leading] == 0) {
+        ++leading;
     }
-    // Allocate enough space in big-endian base58 representation.
-    int size = input.size() * 138 / 100 + 1; // log(256) / log(58), rounded up.
-    std::vector<unsigned char> b58(size);
-    // Process the bytes.
-    while (input.size() > 0) {
-        int carry = input[0];
-        int i = 0;
-        // Apply "b58 = b58 * 256 + ch".
-        for (std::vector<unsigned char>::reverse_iterator it = b58.rbegin(); (carry != 0 || i < length) && (it != b58.rend()); it++, i++) {
-            carry += 256 * (*it);
-            *it = carry % 58;
-            carry /= 58;
-        }
 
-        assert(carry == 0);
-        length = i;
-        input = input.subspan(1);
+    int size = 1 + input.size() * log256_58Ratio / baseScale;
+    std::string result;
+    result.reserve(leading + size);
+    result.assign(leading, '1');
+
+    auto inputAsLongs = ConvertToLongs(input, leading);
+    for (auto i{0U}; i < inputAsLongs.size();) {
+        int64_t remainder{0};
+        for (auto j{i}; j < inputAsLongs.size(); ++j) { // Calculate next digit, dividing inputAsLongs
+            auto accumulator = (remainder << (encodeGroupSize * 8)) | inputAsLongs[j];
```

Comment:
we're using a `|` instead of a `+` since we're shifting by powers of two, so we're adding to zeros